### PR TITLE
test(button): refactor button component

### PR DESCRIPTION
**Pull request by ZhuoyuJin**
### Proposed behaviour

Refactor for Button component to use new cypress-component-react framework for testing.

### Current behaviour

Currently we are using the regular cypress approach with BDD.
### Checklist

- [ ] Commits follow our style guide
- [ ] Related issues linked in commit messages if required
- [ ] Screenshots are included in the PR if useful
- [ ] All themes are supported if required
- [ ] Unit tests added or updated if required
- [ ] Cypress automation tests added or updated if required
- [ ] Storybook added or updated if required
- [ ] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [ ] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

- Run npx cypress open-ct to check if there is newly added test.file
 - Check if the button.test.js file passed
 - Compare all tests were moved/refactored from .feature file -> .test.js file
 - Check if all unnecessary files were deleted, including .json file

### CODE REVIEW

File: src/components/button/button.test.js [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+      }
+    );
+  });
+
+  describe("Check events for Button component", () => {
+    it("should call click function in Actions Tab", () => {
+      const callback = cy.stub();
+
+      CypressMountWithProviders(<Button onClick={callback} />);
+      // then run our tests
```
Comment (DlgSHi) [inline src/components/button/button.test.js:154]: delete this please

File: src/components/button/button.test.js [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            .children()
+            .children()
+            .should("not.have.value", cssValue, "wrap");
+        }
+      }
+    );
+  });
+
+  describe("Check events for Button component", () => {
+    it("should call click function in Actions Tab", () => {
```
Comment (DlgSHi) [inline src/components/button/button.test.js:150]:
```suggestion
    it("should call onClick event", () => {
```

File: src/components/button/button.test.js [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+      const callback = cy.stub();
+
+      CypressMountWithProviders(<Button onClick={callback} />);
+      // then run our tests
+      buttonDataComponent().click({ force: true });
```
Comment (DlgSHi) [inline src/components/button/button.test.js:155]: this is only the action  and no assertion is made for `onClick`

File: src/components/button/button.test.js [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+      buttonDataComponent()
+        .focus()
+        .blur({ force: true })
+        .then(() => {
+          // eslint-disable-next-line no-unused-expressions
+          expect(callback).to.have.been.calledOnce;
+        });
+    });
+
+    it("should be triggered on keyDown", () => {
```
Comment (DlgSHi) [inline src/components/button/button.test.js:172]:
```suggestion
    it("should call the onKeyDown", () => {
```

**Review by DlgSHi — CHANGES_REQUESTED** [review]
1. You have deleted some step that were used in other `feature` files. Please check with steps and move them into `common-steps.js` file.
2. You need to remove `buttonAsSibling.feature` file as well because you covered those tests here.
3. Update the description for the PR :) it is not a `Detail`.

File: src/components/button/button.test.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    );
+
+    it.each([
+      ["after", "10px"],
+      ["before", "10px"],
+    ])(
+      "should set postion to %s for button as a button",
+      (iconPosition, margin) => {
+        CypressMountWithProviders(
+          <Button iconType="add" iconPosition={iconPosition} />
```
[the remark this answers is not in the document]
Comment (ZhuoyuJin) [reply src/components/button/button.test.js:111]: Why do we layout like this here?

File: src/components/button/button.test.js [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+    it.each([
+      ["after", "left"],
+      ["before", "right"],
+    ])(
+      "should set position to %s for icon in a button",
+      (iconPosition, margin) => {
+        CypressMountWithProviders(
+          <Button iconType="add" iconPosition={iconPosition}>
+            {" "}
```
Comment (DlgSHi) [inline src/components/button/button.test.js:78]: why do we need this here ? and the line below ?

Comment (ZhuoyuJin) [reply src/components/button/button.test.js:78]: Added by mistake, will delete it.

**Review by DlgSHi — APPROVED** [review]

File: cypress/support/step-definitions/common-steps.js [hunk 7]
```diff
@@ -22,6 +22,10 @@ import {
 import { dialogTitle } from "../../locators/dialog";
 import DEBUG_FLAG from "..";
 import { pagerSummary } from "../../locators/pager";
+import {
```
Comment (sianford) [inline cypress/support/step-definitions/common-steps.js:25]: Why were Button-specific steps added to this common file? Were they required for other components that had Button embedded?

File: src/components/button/button.test.js [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        );
+
+        icon().should("have.css", `margin-${margin}`, "8px");
+      }
+    );
+
+    it("should render Button with full width", () => {
+      CypressMountWithProviders(<Button fullWidth />);
+
+      buttonDataComponent().should("have.css", "width").and("eq", "1350px");
```
Comment (sianford) [inline src/components/button/button.test.js:89]: Why does the width need to be specified in number of pixels for a full width button? Shouldn't it just be a percentage? I'd have thought that the number of pixels width will differ depending on viewport size.

File: src/components/button/button.test.js [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+      buttonDataComponent()
+        .trigger("keydown", { force: true }, keyCode("rightarrow"))
+        .then(() => {
+          // eslint-disable-next-line no-unused-expressions
+          expect(callback).to.have.been.calledOnce;
+        });
+    });
+  });
+});
```
Comment (sianford) [inline src/components/button/button.test.js:191]: Why are the onChange and onFocus actions not tested here if onClick, onBlur and onKeyDown are?

File: cypress/support/step-definitions/common-steps.js (hunk 7 above)
Comment (ZhuoyuJin) [reply cypress/support/step-definitions/common-steps.js:25]: They are testing for the enable and disable to different types of button, I don't think it's needed for other components.

File: src/components/button/button.test.js [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+      (subtext) => {
+        CypressMountWithProviders(<Button size="large" subtext={subtext} />);
+
+        buttonSubtextPreview().should("have.text", subtext);
+      }
+    );
+
+    it.each(testData)(
+      "should render tooltip message with %s special characters",
+      (tooltipmessage) => {
```
[the remark this answers is not in the document]
Comment (ZhuoyuJin) [reply src/components/button/button.test.js:56]: I think tooltipmessage in this case is a passing in parameter, so I assume if we keep the same name in this function should be ok. But you are right, to make it look consistent, we should change it to 'tooltipMessage'.

File: src/components/button/button.test.js (hunk 9 above)
Comment (ZhuoyuJin) [reply src/components/button/button.test.js:191]: We could do a test for onFocus, but not onChange. onChange is something different with more complicated scenarios. It is hard to test onChange event in button, we discussed it and I decided to leave it for now.

File: cypress/support/step-definitions/common-steps.js (hunk 7 above)
Comment (DlgSHi) [reply cypress/support/step-definitions/common-steps.js:25]: I think those steps are also used in `splitButton` or `multiActionButton`

File: src/components/button/button.test.js (hunk 9 above)
Comment (DlgSHi) [reply src/components/button/button.test.js:191]: I think it should be similar to `onClick` test :)

Comment (ZhuoyuJin) [reply src/components/button/button.test.js:191]: onChange test will fail if I apply the same tests

Comment (sianford) [reply src/components/button/button.test.js:206]: After discussion we've found that onChange does not apply to Button so it's correct that you did not include it.

**Review by sianford — APPROVED** [review]

**Review by Parsium — APPROVED** [review]

**Pull request merged**